File: partiql-parser/src/main/java/org/partiql/parser/SourceLocations.java
package org.partiql.parser;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;
import org.partiql.spi.SourceLocation;

import java.util.Collection;
import java.util.Map;
import java.util.Set;

/**
 * TODO
 */
public class SourceLocations implements Map<String, SourceLocation> {

    private final Map<String, SourceLocation> delegate;

    private SourceLocations(Map<String, SourceLocation> delegate) {
        this.delegate = delegate;
    }

    @NotNull
    @Override
    public Set<Map.Entry<String, SourceLocation>> entrySet() {
        return delegate.entrySet();
    }

    @NotNull
    @Override
    public Set<String> keySet() {
        return delegate.keySet();
    }

    @Override
    public int size() {
        return delegate.size();
    }

    @NotNull
    @Override
    public Collection<SourceLocation> values() {
        return delegate.values();
    }

    @Override
    public boolean containsKey(Object key) {
        return delegate.containsKey(key);
    }

    @Override
    public boolean containsValue(Object value) {
        return delegate.containsValue(value);
    }

    @Override
    public SourceLocation get(Object key) {
        return delegate.get(key);
    }

    @Nullable
    @Override
    public SourceLocation put(String key, SourceLocation value) {
        // TODO: This isn't allowed (yet?)
        return null;
    }

    @Override
    public SourceLocation remove(Object key) {
        // TODO: This isn't allowed (yet?)
        return null;
    }

    @Override
    public void putAll(@NotNull Map<? extends String, ? extends SourceLocation> m) {
        // TODO: This isn't allowed (yet?)
    }

    @Override
    public void clear() {
        // TODO: This isn't allowed (yet?)
    }

    @Override
    public boolean isEmpty() {
        return delegate.isEmpty();
    }

    /**
     * TODO
     */
    public static class Mutable {

        private final Map<String, SourceLocation> delegate = new java.util.HashMap<>();

        /**
         * TODO
         * @param id TODO
         * @param value TODO
         */
        public void set(String id, SourceLocation value) {
            delegate.put(id, value);
        }

        /**
         * TODO
         * @return TODO
         */
        public SourceLocations toMap() {
            return new SourceLocations(delegate);
        }
    }
}

File: partiql-parser/src/main/kotlin/org/partiql/parser/SourceLocations.kt
/*
 * Copyright Amazon.com, Inc. or its affiliates.  All rights reserved.
 *
 * Licensed under the Apache License, Version 2.0 (the "License").
 * You may not use this file except in compliance with the License.
 * A copy of the License is located at:
 *
 *      http://aws.amazon.com/apache2.0/
 *
 * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific
 * language governing permissions and limitations under the License.
 */

package org.partiql.parser

/**
 * Each node is hashable and has a unique identifier. Metadata is kept externally.
 * Delegate once we are on Kotlin 1.7
 */
public class SourceLocations private constructor(
    private val delegate: Map<String, SourceLocation>
) : Map<String, SourceLocation> {

    override val entries: Set<Map.Entry<String, SourceLocation>> = delegate.entries

    override val keys: Set<String> = delegate.keys

    override val size: Int = delegate.size

    override val values: Collection<SourceLocation> = delegate.values

    override fun containsKey(key: String): Boolean = delegate.containsKey(key)

    override fun containsValue(value: SourceLocation): Boolean = delegate.containsValue(value)

    override fun get(key: String): SourceLocation? = delegate[key]

    override fun isEmpty(): Boolean = delegate.isEmpty()

    internal class Mutable {

        private val delegate = mutableMapOf<String, SourceLocation>()

        operator fun set(id: String, value: SourceLocation) = delegate.put(id, value)

        fun toMap() = SourceLocations(delegate)
    }
}
